### PR TITLE
queryx: unset empty values

### LINKED ISSUE #202: using of gocql.UnsetValue
Hi everyone! Recently we faced a following problem. In one of our processes we create a query by specifying ALL columns of a model struct and bind its fields by using method `BindStruct`. Some of the struct fields can be `nil` values and it leads to the problem that we generate ton of tombstones by explicitly sending `null` values to Cassandra. Base driver `gocql` has `gocql.UnsetValue` for unsetting value in statements that are already prepared. Unfortunately, creating custom type and adding method `gocql.MarshalCQL` doesn't solve the problem because `gocql` has different logic for set and unset values: we have to pass `gocql.UnsetValue` directly into `query.Bind` method.

To solve this problem we have some suggestions:
* Add method like `NilAsUnset` for `Queryx` struct. It sets some flag and then method `bindStructArgs` binds `gocql.UnsetValue` for all `nil` fields.
* Add interface with one method `BindCQL() interface{}`. Then in binding process check if value implements this interface. For these fields, instead of themselves, bind a return value of `BindCQL` method. As a result clients have flexibility in defining situations when custom fields should be considered as unset.

What do you think guys? Can we workaround stated problem without adding new functionality? If it's impossible, then we're ready to implement any of the suggested methods by ourselves.

### PR DESCRIPTION
**Pull request by N1cOs**
Resolve #202

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -13,7 +13,7 @@ GOTEST_CPU := 1
 endif
 
 ifndef GOBIN
-GOBIN := $(GOPATH)/bin
+GOBIN := $(shell go env GOPATH)/bin
```
Comment (mmatczuk) [inline Makefile:16]: If missing, let's define GOPATH like GOBIN.

File: example_test.go [hunk 2]
```diff
@@ -106,17 +107,30 @@ func basicCreateAndPopulateKeyspace(t *testing.T, session gocqlx.Session) {
 
 	// Insert song using query builder.
 	insertSong := qb.Insert("examples.songs").
-		Columns("id", "title", "album", "artist", "tags", "data").Query(session)
```
Comment (mmatczuk) [inline example_test.go:109]: Please add new examples if appropriate.

File: transformer.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+package gocqlx
+
+import (
+	"reflect"
+
+	"github.com/gocql/gocql"
+)
+
+// bindTransformer is called right before binding a value to a named parameter.
+type bindTransformer func(name string, val interface{}) interface{}
```
Comment (mmatczuk) [inline transformer.go:14]: Should be exported.

Comment (mmatczuk) [reply transformer.go:14]: I think we can leave is as `Transformer`, it looks generic enough to use it with parsing rows if needed.

File: transformer.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+// UnsetEmpty unsets all empty parameters.
+// It helps to avoid tombstones when using the same insert/update
+// statement for filled and partially filled parameters.
+var UnsetEmpty = func(name string, val interface{}) interface{} {
```
Comment (mmatczuk) [inline transformer.go:19]: It should be more explicit that this is a transformer.

File: queryx.go [hunk 5]
```diff
@@ -136,7 +144,7 @@ func (q *Queryx) BindStructMap(arg0 interface{}, arg1 map[string]interface{}) *Q
 	return q
 }
 
-func bindStructArgs(names []string, arg0 interface{}, arg1 map[string]interface{}, m *reflectx.Mapper) ([]interface{}, error) {
+func bindStructArgs(names []string, arg0 interface{}, arg1 map[string]interface{}, m *reflectx.Mapper, tr bindTransformer) ([]interface{}, error) {
```
Comment (mmatczuk) [inline queryx.go:147]: It already has 2 Query fields passed in let's make it a method.

File: transformer.go [hunk 6]
```diff
@@ -0,0 +1,13 @@
+// Copyright (C) 2017 ScyllaDB
+// Use of this source code is governed by a ALv2-style
+// license that can be found in the LICENSE file.
+
+package gocqlx
+
+// Transformer transforms the value of the named parameter to another value.
+type Transformer func(name string, val interface{}) interface{}
+
+// DefaultBindTransformer just do nothing.
```
Comment (N1cOs) [inline transformer.go:16]: It sounds a bit strange but I couldn't come up with anything better :)

File: Makefile (hunk 1 above)
Comment (N1cOs) [reply Makefile]: Done

File: example_test.go (hunk 2 above)
Comment (N1cOs) [reply example_test.go:109]: I remained your insert statement, just save it in the separate variable. I suppose it's a good example how one statement can be used in situations with all filled params and partially filled.

If you think it's inappropriate in this example, I can remove it.

File: transformer.go (hunk 3 above)
Comment (N1cOs) [reply transformer.go]: Named it as `Transformer`

File: transformer.go (hunk 4 above)
Comment (N1cOs) [reply transformer.go]: Initially I though it would be some redundancy in word transformer but ok

File: queryx.go (hunk 5 above)
Comment (N1cOs) [reply queryx.go]: Done

File: transformer.go (hunk 6 above)
Comment (N1cOs) [reply transformer.go:16]: I set it as nil to save compatibility with older versions. Maybe someone uses insert/update statements with nil values to explicitly (or implicitly, who knows) delete columns

File: example_test.go (hunk 2 above)
Comment (mmatczuk) [reply example_test.go:109]: It break the convention, feel free to copy the example and fully adjust it to present your usecase.

**Comment by mmatczuk** [comment]
All is good let's fix the example. The current example is called `basicCreateAndPopulateKeyspace` please add something proper so that folks can grasp the idea.

File: example_test.go (hunk 2 above)
Comment (N1cOs) [reply example_test.go]: Ok, added separate example

**Review by mmatczuk — APPROVED** [review]

**Comment by mmatczuk** [comment]
Thanks for the contribution!

**Comment by N1cOs** [comment]
@mmatczuk can we tag a new release, so we can update our dependencies?

**Comment by mmatczuk** [comment]
Let's do that.

**Comment by mmatczuk** [comment]
Here it is @N1cOs https://github.com/scylladb/gocqlx/releases/tag/v2.6.0

**Pull request merged**